text/3857-cfg-version.md

```diff
@@ -0,0 +1,609 @@
+[drawbacks]: #drawbacks
+
+People may be using `--cfg rust` already and would be broken by this change.
+There are no compatibility concerns with predicate names.
+
+This does not include a solution for adopting this within `Cargo.toml` without waiting for an MSRV bump.
+
+Traditionally, maintainers only test their MSRV and latest, assuming those will catch every issue.
+While that isn't always true today (e.g. some Cargo features go from "unknown" warning to "unstable" error to supported and MSRV might be in the warning phase),
+having distinct implementations for different Rust versions can make the testing matrix more complex.
+Tools like [`cargo hack`](https://crates.io/crates/cargo-hack) can help which can run commands on not just one toolchain version but also the every version starting with the MSRV with a command like `cargo hack --rust-version --version-step 1 check`.
+
+This does not help with probes for specific implementations of nightly features which still requires having a `build.rs`.
+
+Libraries could having ticking time bombs that accidentally break or have undesired behavior for some future Rust version that can't be found until we hit that version.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## `since` cfg predicate
+
+We could offer a `before` operator but that is already covered by `not(since)`.
+
+The `since` name was taken from 
+[rustversion](https://crates.io/crates/rustversion) and the `#[deprecated(since)]` / `#[stable(since)]` attributes.
+This better conveys what operation is being performed than the original `version` name
+and leaves room for related predicates like `before`.
+We could also call this `minimum`, or support comparison operators in the spirit of [RFC 3796](https://github.com/rust-lang/rfcs/pull/3796).
+The risk with a general word like `since` is if we gain support for other data types in cfgs, like integers for embedded development.
+The name `since` might apply in some situations but not others and its unclear if we'd want to generalize it past versions.
+While having a specific name avoids these concerns.
+
+We could swap the order of parameters and make `rust` a default for the second parameter to allow `#[cfg(since("1.95"))]` as a shorthand.
+However, this would look confusing in Cargo and doesn't seem like its offering enough of a benefit to be worth the costs.
+
+The `ConfigurationVersion` is sloppy with the string literal's syntax (relying on `--check-cfg`) so that
+- Allows evolution without requiring an MSRV bump
+- Its consistent with other predicates, e.g. `#[cfg(foo = "1.0")]`
+
+If we were stricter on the syntax,
+we could allow for version numbers to be directly accepted, without quotes 
+(e.g. `#[cfg(since(rust, 1.95.0))]`).
+If we ever decided to support operators (e.g.`#[cfg(since(rust, "=1.95.0"))]`, see `--check-cfg`), then we'd need to decide if those also go outside the string or then require a string, being inconsistent.
```

Comment:
This also seems to cause problems when it is combined with the "different versions may have different `cmp` implementations and we may want to account for that by allowing setting a policy" idea, doesn't it?

Then we might be expected to allow for non-SemVer strings in this "bare words" form. Then the parser might have to reason about how to properly consume the tokens instead of just extracting a string, possibly according to a policy which, in the current proposal, would be specified *after* that sequence of tokens? Or we just disallow even parsing those, which seems like it would have other problems.

Reply:
Included this in 3a7609a4